### DISCUSSION TITLE
StrongSwan server itself in tunnel and its strange behavior despite passthrough policy

### DISCUSSION THREAD

**Discussion by TomHsiung** [strongswan/strongswan#1628] [General]
Hello, everyone

I have a StrongSwan server and it provides both roadwarrior (conn R1, R2) and site2site (s2s1) connection service. When traffic via R1 is unencrypted, they would be directed into s2s1, while traffic handled via R2 is just dealt by gateway iptables rules. So I set up some passthrough policy to bypass the s2s1 for traffic targeting some IPs (most importantly, the StrongSwan server itself, so I could remotely connect it via SSH through R1 tunnel). It worked very well until recently.

Recently, I added the IP address of the StorngSwan server (it has two interfaces, one is the IP address behind the other which acts as NAT gateway) so the traffic derived from the StrongSwan server could go through the s2s1 tunnel. But today I find that I am no longer able to connect the server via R1 tunnel anymore. The strangest thing is, if I close the s2s1 tunnel after I established R1 tunnel, the communication between my client and the StrongSwan server is normal. Later if I turn on the s2s1 again while I keep the status of the R1 tunnel the communication is still normal.

Communication with other hosts which are on the same subnet of the StrongSwan server is normal. I don't know if there is something wrong with the passthrough policy in `ipsec.conf` or it is just a bug.

**Comment by tobiasbrunner** [comment]
> Recently, I added the IP address of the StorngSwan server (it has two interfaces, one is the IP address behind the other which acts as NAT gateway) so the traffic derived from the StrongSwan server could go through the s2s1 tunnel.

I don't understand what you mean by that.

> But today I find that I am no longer able to connect the server via R1 tunnel anymore.

What does that mean exactly? Connect what?

Also, please provide the information requested in #196 (and anything beyond that that applies to your situation).

**Comment by TomHsiung** [comment]
OK. Please let me summarize my experience with the event briefly.

Ubuntu 22.04 TLS and StrongSwan 5.9.5. The machine has two ethernets, one with IP `A.A.A.A` and the other `B.B.B.B`. StrongSwan is listening on both of the IP addresses but only `A.A.A.A` is used for VPN.

And below is my step-by-step experienced.

#1 Establish site-to-site VPN connection
`ESTABLISHED xx minutes ago, A.A.A.A[domainA]...remoteIP[remoteDomain]`

site-to-site tunnel (tunnel1)
`X.X.X.X/X C.C.C.C/C B.B.B.B/32  === 0.0.0.0/0`

ipsec.conf passthrough
`conn passthrough
        left=127.0.0.1
        leftsubnet=C.C.C.C/C
        rightsubnet=B.B.B.B/32
        type=passthrough
        authby=never
        auto=route`

#2 Establish roadwarrior VPN connection from a Mac client to domainA
`ESTABLISHED xx minutes ago, A.A.A.A[domainA]...clientIP[clientIdentity]`

rightsubnet
`C.C.C.C/C`

road warrior tunnel
`C.C.C.C/C  === 0.0.0.0/0`

Here, the issue turned up, pings from client to B.B.B.B timed out. But pings from B.B.B.B to C.C.C.C is normal.

#3 Step 3
`ipsec down tunnel1`

Then, pings from client to B.B.B.B became normal.

#4 Step 4
`ipsec up tunnel1` and site-to-site tunnel reestablished

Finally, pings from client to B.B.B.B was still normal.

#5 Step 5
On the Mac client, disconnect the roadwarrior connection before reconnecting it again. Problem in step 2 reoccurred after the reconnection.

**Reply by tobiasbrunner** [reply]
You again only provide partial information. Is `C.C.C.C/C` reachable from `B.B.B.B/32` without the roadwarrior connection? Do you actually want that traffic to be unprotected?

> `leftsubnet=C.C.C.C/C rightsubnet=B.B.B.B/32`

Hm, these seem reversed. From the server's point of view the remote addresses (`rightsubnet`) are in `C.C.C.C/C` while `B.B.B.B/32` is a local address (`leftsubnet`). For the policies it doesn't really matter, however, there is also a `throw` route installed in table 220 together with the outbound policy (i.e. to the remote subnet, in this case to `B.B.B.B/32`, instead of the actual remote subnet `C.C.C.C/C`). Check the routing tables and policies to see how that actually looks.

But be aware that the bypass policy applies to both connections, you can't selectively install one that only applies to the site-to-site tunnel (unless there are other measures like using custom priorities or marks).

**Comment by TomHsiung** [comment]
C.C.C.C/C is the virtual IP by the roadwarrior connection, so if without the roadwarrior connection, C.C.C.C/C does not exist.

> You again only provide partial information. Is C.C.C.C/C reachable from B.B.B.B/32 without the roadwarrior connection?

No, I want packets to be encrypted.

> Do you actually want that traffic to be unprotected?

Yep, the passthrough policy (`rightsubnet=C.C.C.C/C leftsubnet=B.B.B.B/32`) will affect the roadwarrior connection too (the roadwarrior connection between C.C.C.C/C and B.B.B.B/32 will be bypassed so it would be impossible to make an end-to-end communication via the roadwarrior connection between the two ends)

> Hm, these seem reversed.

> you can't selectively install one that only applies to the site-to-site tunnel

**Reply by tobiasbrunner** [reply]
> C.C.C.C/C is the virtual IP by the roadwarrior connection, so if without the roadwarrior connection, C.C.C.C/C does not exist.

So you configured `C.C.C.C/C` in `rightsourceip`? Make sure you don't configure `rightsubnet` on such a roadwarrior connection.



> No, I want packets to be encrypted.

Then why install a bypass policy at all? That doesn't make sense to me.

> the roadwarrior connection between C.C.C.C/C and B.B.B.B/32 will be bypassed so it would be impossible to make an end-to-end communication via the roadwarrior connection between the two ends

What do you mean?

**Reply by TomHsiung** [reply]
Because the site-to-site policy is `X.X.X.X/X C.C.C.C/C B.B.B.B/32  === 0.0.0.0/0`, if I don't let the packets from B.B.B.B/32 to C.C.C.C/C bypass the site-to-site tunnel, they would go into the site-to-site tunnel. The site-to-site tunnel is connected to another remote device, not those ones with virtual IP C.C.C.C/C
> Then why install a bypass policy at all? That doesn't make sense to me.

The flow is host1 connection B.B.B.B/32 via roadwarrior (and gets a virtual IP C.C.C.C/C), then all packets from host1 are tunneled to B.B.B.B/32, decrypted. Later, those traffic which fulfill the site-to-site policy would be tunneled to a remote host device for further purpose.

**Reply by tobiasbrunner** [reply]
Why `C.C.C.C/C` shouldn't it be `C.C.C.C/32` if it's a virtual IP assigned to the roadwarrior? And yes, that could actually result in an outbound policy that has the same priority as one of the outbound policies for the site-to-site tunnel as both have a `/32` on one end and a `/0` on the other  (one is for `0.0.0.0/0 === C.C.C.C/32` and the other is for `B.B.B.B/32 === 0.0.0.0/0`). You could perhaps avoid that by using `B.B.B.B/31` for the site-to-site tunnel. Or you set a custom priority for the roadwarrior connection so that it's preferred over the other.

Again, a bypass policy won't help you as it will exclude traffic also from the roadwarrior connection.

**Reply by TomHsiung** [reply]
Thanks. I am considering using iptable rules to fix this issue.

**Reply by tobiasbrunner** [reply]
With marks? Or what do you mean?

**Reply by TomHsiung** [reply]
Since I must exclude the `B.B.B.B/32` from the site-to-site policy, I am considering transferring B.B.B.B/32 to another IP may be D.D.D.D/32 which might be put into the site-to-site policy. Actually, I want packets generated by the strongswan server to go into the site-to-site tunnel, i.e. for Ubuntu upgrading, DNS querying (the strongswan server also runs named.service)

**Comment by TomHsiung** [comment]
However, how do you explain that the communication became fine after `ipsec down tunnel1` (site-to-site), and was still fine with additional re-establishment via `ipsec up tunnel1` (site-to-site)?
> Again, a bypass policy won't help you as it will exclude traffic also from the roadwarrior connection.

**Reply by tobiasbrunner** [reply]
Because the kernel stores the policies in a list (or multiple nowadays) ordered by priorities and if multiple have the same priority which one is used depends on the order in which they are added/removed.

**Reply by TomHsiung** [reply]
Could you please guide me how to do this? Thanks

> Or you set a custom priority for the roadwarrior connection so that it's preferred over the other.

**Comment by TomHsiung** [comment]
And one thing I don't understand is pings from C.C.C.C/C to devices within the same subnet of the StrongSwan server B.B.B.B/32 (i.e., B.B.B.254/32) is fine, provided the passthrough policy below:

`conn passthrough left=127.0.0.1 leftsubnet=C.C.C.C/C rightsubnet=B.B.B.0/24 type=passthrough authby=never auto=route`

**Reply by tobiasbrunner** [reply]
For such forwarded traffic you actually need a bypass policy as you want to avoid that it's sent through the site-to-site tunnel (for traffic that's sent to addresses local to the gateway that's not necessary). And the response traffic for those addresses won't conflict with the site-to-site tunnel as that only lists the one IP address from the B subnet that's assigned to the gateway (only for that could there be conflicting outbound policies).

**Reply by TomHsiung** [reply]
Thanks for the response. However, I think I had not described my scenario correctly. I will update my scenario later.

**Comment by TomHsiung** [comment]
Passthrough policy

> conn passthrough2
>         left=127.0.0.1
>         leftsubnet=10.0.12.0/24
>         rightsubnet=10.0.0.0/24
>         type=passthrough
>         authby=never
>         auto=route

Roadwarrior tunnel policy

> leftsubnet=0.0.0.0/0
> rightsourceip=10.0.12.0/24

And this site2site tunnel rules

> leftsubnet=10.0.0.96/28,10.0.12.0/24,10.0.0.1/32
> rightsubnet=0.0.0.0/0

pings from 10.0.12.1 to 10.0.0.101/32 can get echo response normally. However, pings from 10.0.12.1 to 10.0.0.1/32 timeout. Client with 10.0.12.1 is remote away and it connected to the 10.0.0.1 via the roadwarrior tunnel. Site-to-site tunnel is established to another remote device.

**Reply by tobiasbrunner** [reply]
If you look at the installed policies, you should see what the problem is. It's due to the [previously discussed](https://github.com/strongswan/strongswan/discussions/1628#discussioncomment-5454317) conflict between the site-to-site tunnel's outbound policy that has `10.0.0.1/32` as source address and `0.0.0.0/0` as destination and the roadwarrior connection's outbound policy that has `0.0.0.0/0` as source and `10.0.12.1/32` as destination address. These have the same priority so which one is used depends on the installation order. As mentioned before, either change the automatic priority by using a different local subnet for the site-to-site connection, or manually change the priority for the connections (or at least one of them). Note that the latter is only possible via swanctl.conf, not ipsec.conf.

**Reply by TomHsiung** [reply]
However, the device with IP `10.0.0.101/32` is also include in the site-to-site tunnel (`10.0.0.96/28`). What is the reason why the two tunnel policies do not conflict for `10.0.0.101/32`?

> leftsubnet=10.0.0.96/28,10.0.12.0/24,10.0.0.1/32
> rightsubnet=0.0.0.0/0

**Reply by tobiasbrunner** [reply]
The priorities are calculated based on the subnet sizes of the selectors. And the one for the outbound policy with source `10.0.0.96/28` is obviously lower than the one with `10.0.0.1/32` and, therefore, won't conflict with the roadwarrior policy that has a `/32` on one end as well.

**Reply by TomHsiung** [reply]
Wow! Fantastic.

> The priorities are calculated based on the subnet sizes of the selectors.

Is there any reference to this "rule of priorities". Or could you tell me deeper about it? Thank you.

**Reply by tobiasbrunner** [reply]
See [here](https://github.com/strongswan/strongswan/blob/5284cecddc4dbbbb0772313190af3aa690278123/src/libcharon/plugins/kernel_netlink/kernel_netlink_ipsec.c#L713-L754).

**Reply by TomHsiung** [reply]
Thank you. Now I know the size of the subnets defined by the tunnel policy will affect the priority of that policy.

**Comment by TomHsiung** [comment]
Finally, your recommendation works. Thank you.

> You could perhaps avoid that by using B.B.B.B/31 for the site-to-site tunnel.
